django/thunderstore/api/cyberstorm/views/community.py

```diff
@@ -0,0 +1,214 @@
+from thunderstore.community.models import Community
+
+
+class CommunityAPIView(APIView):
+    permission_classes = []
+
+    @swagger_auto_schema(
+        responses={200: CyberstormCommunitySerializer()},
+        operation_id="api.community",
+    )
+    def get(self, request: HttpRequest, community_identifier: str) -> HttpResponse:
+        # TODO: Filter out the common shared packages somehow
+        c_q = Community.objects.annotate(
+            pkgs=Count("package_listings", distinct=True),
+            downloads=Sum(
+                "package_listings__package__versions__downloads", distinct=True
+            ),
+        )
+        try:
+            c = c_q.get(
+                identifier=community_identifier,
+            )
+        except Community.DoesNotExist:
+            raise Http404(
+                f"Coudln't find community with the identifier {community_identifier}"
+            )
+
+        serializer = self.serialize_results(c)
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def serialize_results(self, c: Community):
+        return CyberstormCommunitySerializer(
+            {
+                "name": c.name,
+                "namespace": c.identifier,
+                "download_count": c.downloads,
+                "package_count": c.pkgs,
+                "image_source": c.background_image_url,
+                "server_count": 0,
+                "description": c.description,
+                "discord_link": c.discord_url,
+            }
+        )
+
+
+class CommunitiesAPIView(APIView):
+    permission_classes = []
+
+    @swagger_auto_schema(
+        responses={200: CommunitiesListCyberstormSerializer()},
+        operation_id="api.communities",
+    )
+    def get(
+        self,
+        request: HttpRequest,
+    ) -> HttpResponse:
+
+        qp = CommunitiesQueryParameterCyberstormSerializer(data=request.query_params)
+        if not qp.is_valid():
+            return Response(qp.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        params: OrderedDict = qp.validated_data
+        communities_qs = self.get_queryset(params)
+        communities_qs = self.order_queryset(params["ordering"], communities_qs)
+        communities_page = self.paginate(params, communities_qs, params["page_size"])
+        communities_count = len(communities_qs)
+        serializer = self.serialize_results(
+            communities_page, communities_count, params["page_size"]
+        )
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def get_full_cache_vary(self, params: OrderedDict) -> str:
+        cache_vary = "communities_cache_key"
+        cache_vary += f".{params['ordering']}"
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{params['page_size']}"
+        return cache_vary
+
+    def get_queryset(self, params=None) -> QuerySet[Community]:
+        return Community.objects.filter(is_listed=True,).annotate(
+            pkgs=Count(
```

Comment:
Don't quite recall how well this is supported, but if some of these subqueries could be moved to variables, it could improve the readability of the code.

Reply:
I'm not sure if those can be, but I can try it out for sure

Reply:
Readability wasn't improved much from putting in separate subquery variable. I'll keep it as is for now.